test/adainterop/main.c

```diff
@@ -0,0 +1,123 @@
+// This file is part of the adac-hybrid-msp430 distribution
+// Copyright (C) 2020, European Space Agency
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <Fpu/Fpu.h>
+
+// Reporting functions defined in Ada "Reporting" module
+void reporting__init();
+void reporting__reportsuccess();
+void reporting__reporterror(const int32_t errno);
+
+// Tested functions defined in "AdaModule" module, but exported to C
+int32_t addInt32(const int32_t a, const int32_t b);
+int16_t addInt16(const int16_t a, const int16_t b);
+float addFloat(const float a, const float b);
+float addInt16AndFloat(const int16_t i, const float f);
+int32_t addMany(const int32_t a1, const int32_t a2, const int32_t a3, const int32_t a4);
+
+static bool test_addInt32()
+{
+    const int32_t a = 16;
+    const int32_t b = 52;
+    const int32_t result = addInt32(a, b);
+    if (result == (a + b))
+        return true;
+    return false;
+}
+
+static bool test_addInt16()
+{
+    const int16_t a = 16;
+    const int16_t b = 52;
+    const int16_t result = addInt16(a, b);
+    if (result == (a + b))
+        return true;
+    return false;
+}
+
+static bool test_addFloat()
+{
+    const float a = 16.5;
+    const float b = 52.25;
+    const float result = addFloat(a, b);
+    if (result == (a + b))
+        return true;
+    return false;
+}
+
+static bool test_addInt16AndFloat()
+{
+    const int16_t i = 16;
+    const float f = 52.25;
+    const float result = addFloat(i, f);
+    if (result == (i + f))
+        return true;
+    return false;
+}
+
+static bool test_addMany()
+{
+    const int32_t a1 = 1234;
+    const int32_t a2 = 64561;
+    const int32_t a3 = 15632;
+    const int32_t a4 = 52345;
+    const int32_t result = addMany(a1, a2, a3, a4);
+    if (result == (a1 + a2 + a3 + a4))
+        return true;
+    return false;
+}
+
+void runTests(void)
```

Comment:
From the perspective of testing: corner cases and equivalence classes:
0 + 0
0 + -10
0 + 10
-10 + 0
10 + 0
-10 + 20
20 + -10
maxint + maxint
maxint + 0
0 + maxint
minint + minint
minint + 0
0 + minint
minint + minint
maxint + minint
minint + maxint

Reply:
The purpose of this is not to test the actual arithmetic, but handling of integers of different sizes. This actually detected issues with 16-bit ints on MSP430.